include/miopen/miopen.h

```diff
@@ -4701,6 +4701,106 @@ miopenReduceTensor(miopenHandle_t handle,
+    miopenTensorConvolutionY = 3,
+} miopenTensorName_t;
+
+typedef enum
+{
+    miopenSearchResultsOrderByTime   = 0,
+    miopenSearchResultsOrderByMemory = 1,
+} miopenSearchResultsOrder_t;
+
+miopenStatus_t miopenCreateConvProblem(miopenProblem_t* problem,
+                                       miopenConvolutionDescriptor_t operatorDesc,
+                                       miopenProblemDirection_t direction);
+
+miopenStatus_t miopenDestroyProblem(miopenProblem_t problem);
+
+// Universal problem should have named tensor fields rather than ordered.
+// Previously, different directions had different input orders, which may lead to confusion.
+// Implicitly ordered fields would also lead to a whole lot of magical ids all over the internals.
+miopenStatus_t miopenSetProblemTensorDescriptor(miopenProblem_t problem,
+                                                miopenTensorName_t name,
+                                                const miopenTensorDescriptor_t descriptor);
+
+/**
+ * The miopenSearchOptions will allow the user to configure how search will be
+ * used. This can include things like exhaustiveSearch, and ordering. In the
+ * future, this can support setting different find mode options.
+ */
+MIOPEN_DECLARE_OBJECT(miopenSearchOptions);
+
+miopenStatus_t miopenCreateSearchOptions(miopenSearchOptions_t* options);
+
+miopenStatus_t miopenDestroySearchOptions(miopenSearchOptions_t options);
+
+miopenStatus_t miopenSetSearchOptionTuning(miopenSearchOptions_t options, int value);
+
+miopenStatus_t miopenSetSearchOptionResultsOrder(miopenSearchOptions_t options,
+                                                 miopenSearchResultsOrder_t value);
+
+miopenStatus_t miopenSetSearchOptionWorkspaceLimit(miopenSearchOptions_t options, size_t value);
+
+MIOPEN_DECLARE_OBJECT(miopenSolution);
+
+/**
+ * Finds solutions to a problem by running different applicable solutions. Memory is automatically
+ * allocated.
+ */
+miopenStatus_t miopenFindSolutions(miopenHandle_t handle,
+                                   miopenProblem_t problem,
+                                   miopenSearchOptions_t options,
+                                   miopenSolution_t* solutions,
+                                   size_t* numSolutions,
+                                   size_t maxSolutions);
+
+/**
+ * Runs the solution using the passed in buffers. Corresponding tensor
+ * descriptors are passed in so if a user would like to dynamically change
+ * the sizes, they can do so. If dynamic changes are not supported(or are
+ * invalid), we should return an error to the user.
+ */
+miopenStatus_t miopenRunSolution(miopenHandle_t handle,
+                                 miopenSolution_t solution,
+                                 size_t nInputs,
+                                 const miopenTensorName_t* names,
+                                 const miopenTensorDescriptor_t* descriptors,
+                                 void* const* buffers,
```

Comment:
The signature seems redundant. The user wants to run a Solution for some Problem. The Problem "knows" all TensorDescriptors as well as a number of tensors/buffers. So if the instance of a Problem still exists, the user can pass Problem and a set of {name, buffer_ptr} pairs.

Reply:
Tensor descriptors are optional here. That's not my idea originally, and a lot of solvers won't support mismatching tensor descriptors right now.

Reply:
We added the tensor descriptors to support dynamic tensors in the future, but maybe this doesn't really make sense.

Reply:
Actually, there are solvers which support it. What we need is a way to return that information to the user and to properly check for that before executing.